cassiopeia/data/CassiopeiaTree.py

```diff
@@ -2046,6 +2061,564 @@ def get_unmutated_characters_along_edge(
+                            child_possible_states.append(c)
+                    # '*' represents all non-missing states (including 0), and
+                    # is a possible state when all children are missing, as any
+                    # state could have occurred at the parent if all missing
+                    # data events occurred independently.
+                    if child_possible_states == []:
+                        possible_states[char] = [
+                            "*",
+                            self.missing_state_indicator,
+                        ]
+                    else:
+                        possible_states[char] = list(
+                            set.intersection(*child_possible_states)
+                        )
+                        if 0 not in possible_states[char]:
+                            possible_states[char].append(0)
+
+            likelihoods_at_n = [{} for i in range(self.n_character)]
+
+            # We calculate the likelihood of the states at each character at
+            # the current node according to the recurrence relation
+            for char in range(self.n_character):
+                for s in possible_states[char]:
+                    likelihood_s = 1
+                    for child in self.children(n):
+                        likelihood_for_child = 0
+                        for s_ in likelihoods_at_nodes[child][char]:
+                            likelihood_at_state_for_child = (
+                                transition_probability(
+                                    s,
+                                    s_,
+                                    self.get_branch_length(n, child),
+                                    char,
+                                )
+                                * likelihoods_at_nodes[child][char][s_]
+                            )
+                            # Here we take into account the probability of
+                            # stochastic missing data
+                            if self.is_leaf(child):
+                                if s_ == -1 and s != -1:
+                                    likelihood_at_state_for_child += (
+                                        1
+                                        - missing_probability_function_of_time(
+                                            heritable_missing_rate,
+                                            self.get_branch_length(n, child),
+                                        )
+                                    ) * stochastic_missing_rate
+                                if s_ != -1:
+                                    likelihood_at_state_for_child *= (
+                                        1 - stochastic_missing_rate
+                                    )
+                            likelihood_for_child += (
+                                likelihood_at_state_for_child
+                            )
+                        likelihood_s *= likelihood_for_child
```

Comment:
Work with log-likelihoods instead, and use `scipy.special.logsumexp` to perform marginalization.